### ISSUE TITLE
remove_session_cb is not called at tls_process_new_session_ticket

### ISSUE THREAD

**Issue by moyaldror** [openssl/openssl#6377]
Hi,
In commit [36ff232cf2bf5dfcaf9e60a8c492439428a243bb](https://github.com/openssl/openssl/commit/36ff232cf2bf5dfcaf9e60a8c492439428a243bb), you removed the lines below from `tls_process_new_session_ticket` function:
```
       if (i & SSL_SESS_CACHE_CLIENT) {	
           /*	
            * Remove the old session from the cache. We carry on if this fails	
            */	
           SSL_CTX_remove_session(s->session_ctx, s->session);	
       }
```	
If understand the changes correctly, it was entered to support the case that multiple tickets are being received from the server, and you don't want each ticket processing to remove from the cache the session that was a dup result from the previous ticket.
The thing is, as it is recommended (and in 0-RTT kind of mandatory to support anti-replay) to use a ticket only once, but after this commit, receiving a NewSessionTicket will never remove previous sessions from the cache and those sessions (which are not relevant anymore) will be removed only at timeout.

Why not add `recieved_tickets` to ssl_st (just like you already added `sent_tickets`) and re-enter the code that was removed (from above) to this new piece:
```
       if (i & SSL_SESS_CACHE_CLIENT && SSL_IS_TLS13(s) && s->recieved_tickets== 0) {	
           /*	
            * Remove the old session from the cache. We carry on if this fails	
            */	
           SSL_CTX_remove_session(s->session_ctx, s->session);	
       }
```

**Comment by mattcaswell** [comment]
We could do that, which gives some functionality a bit more like what happens in TLSv1.2. But in TLSv1.2 it was considered ok to use a ticket more than once, and a new session ticket is likely not to be issued until the previous one is considered "too old". So, in that context, it makes sense to remove the old ticket when you get a new one. But for TLSv1.3 where tickets are encouraged to be single use it just doesn't seem to make as much sense any more. The arrival of a new ticket generally should not be considered as signalling that a previous one is too old. It should be perfectly ok to use both the newly arrived ticket and the previous one.

It makes even less sense in the context of multi-tickets. If you receive 2 tickets each time from the server. Why should you only remove one of them when you get new ones?

Arguably we should remove a session from the client cache when we attempt to *use* it. But this is completely independent of the arrival of new tickets.

**Comment by moyaldror** [comment]
Why in multi-tickets should you remove both sessions? they will derive different PSK values as each should have a distinct nonce so i see no reason to remove 2 entries.
BTW - Have you missed this when adding multi-ticket support?
https://github.com/openssl/openssl/blob/c22365b399f62af4a81e9202500cd2cbd9c23a9d/ssl/statem/statem_srvr.c#L3809-L3814

Back to the point - So maybe cache delete should be done for sessions with tls versions older than TLSv1.3 and for TLSv1.3 with 0-RTT?
At this point are able to know if the client has tried has sent (or at least tried) 0-RTT data.

And thanks for the quick reply!

**Comment by mattcaswell** [comment]
> BTW - Have you missed this when adding multi-ticket support?

Nice catch! Yes, it was missed. I raised #6387 to track this issue.

> So maybe cache delete should be done for sessions with tls versions older than TLSv1.3 and for TLSv1.3 with 0-RTT?

Yes, that seems reasonable (although in the 0-RTT case it should be done at the point of use, not on receipt of a new ticket).

**Comment by moyaldror** [comment]
> Nice catch! Yes, it was missed. I raised #6387 to track this issue.

Glad to be helpful :)

> although in the 0-RTT case it should be done at the point of use, not on receipt of a new ticket

I don't like this approach for the same reason I don't like the anti-replay that was added lately - it's naive and very strict.
When extracting a session from the cache, only application layer knows if the session is about to send 0-RTT data, so at this point, the only thing that can be done is to try and guess by asking the session if max_early > 0 (which is again very strict in my opinion).

**Comment by mattcaswell** [comment]
> I don't like this approach for the same reason I don't like the anti-replay that was added lately - it's naive and very strict.
When extracting a session from the cache, only application layer knows if the session is about to send 0-RTT data, so at this point, the only thing that can be done is to try and guess by asking the session if max_early > 0 (which is again very strict in my opinion).

I suppose we could remove the session from the cache at the point that the client attempts to send early data. Or alternatively at the point that the server accepts it.

WRT anti-replay I think being strict by default is probably the right thing to do (prevent as many people as possible from shooting themselves in the foot). That doesn't mean I'm averse to the possibility of having an option to turn that feature off. It's entirely possible that an application can manage the replay risks in a different way that is unknown to OpenSSL.

**Comment by mattcaswell** [comment]
I raised issue #6389 for the anti-replay issue.

**Comment by moyaldror** [comment]
The issues addressed in [#6389](https://github.com/openssl/openssl/issues/6389) are exactly what bothered me! Thanks!

**Comment by mattcaswell** [comment]
Fix in #6601.
